File: iOS/Chill With Me/Models/CurrentStatus.swift
//
//  CurrentStatus.swift
//  Chill With Me
//
//  Created by Phillip Hoang on 2018-09-17.
//  Copyright © 2018 IC Heart Technologies. All rights reserved.
//

import Foundation

let R2C = "R2C"
let NA = "NA"
let DND = "DND"

class CurrentStatus: Codable {
    
    var statusType: String
    var statusDescription: String
    var timeUntil: String
    var refreshToken: String
    
    init(type: String, description: String, time: String) {
        self.statusType = type
        self.statusDescription = description
        self.timeUntil = time
        self.refreshToken = ""
    }
    
    init() {
        self.statusType = NA
        self.statusDescription = "Not available"
        self.timeUntil = "0000000000"
        self.refreshToken = ""
    }
    
    func isTimeValid() -> Bool {
        if let numericTime = Double(self.timeUntil) {
            if numericTime > NSDate().timeIntervalSince1970 {
                return true
            } else {
                return false
            }
        } else {
            return false
        }
    }
    
    func isR2C() -> Bool {
        if self.isTimeValid() {
            if self.statusType == R2C {
                return true
            } else {
                return false
            }
        } else {
            return false
        }
    }
    
    func isBusy() -> Bool {
        if self.isTimeValid() {
            if self.statusType == DND {
                return true
            } else {
                return false
            }
        } else {
            return false
        }
    }
    
    func isBusyOrR2C() -> Bool {
        if self.isR2C() || self.isBusy() {
            return true
        } else {
            return false
        }
    }
}

File: Android/java/r2chill/Model/CurrentStatus.java
package example.r2chill.Model;

public class CurrentStatus {
    private String statusType;
    private String statusDescription;
    private String timeUntil;
    private String refreshToken;

    private String R2CStatus = "R2C";
    private String NA = "NA";
    private String DND = "DND";

    public CurrentStatus(String id, String description, String time) {
        this.statusType = id;
        this.statusDescription = description;
        this.timeUntil = time;
        this.refreshToken = "";
    }

    // Blank status constructor
    public CurrentStatus() {
        this.statusType = NA;
        this.statusDescription = "Not available";
        this.timeUntil = "0000000000";
        this.refreshToken = "";
    }

    public boolean isBusyOrR2C() {
        String statusType = getStatusType();

        if (statusType.equals("R2C") || statusType.equals("DND")) {
            long time = System.currentTimeMillis() / 1000;
            long statusTime = Long.parseLong(this.getTimeUntil());
            if (statusTime > time) {
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    public boolean isR2C() {
        if (this.statusType.equals("R2C")) {
            long time = System.currentTimeMillis() / 1000;
            long statusTime = Long.parseLong(this.getTimeUntil());
            if (statusTime > time) {
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    public boolean isBusy() {
        if (this.statusType.equals("DND")) {
            long time = System.currentTimeMillis() / 1000;
            long statusTime = Long.parseLong(this.getTimeUntil());
            if (statusTime > time) {
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    public String getRefreshToken() {
        return this.refreshToken;
    }

    public void setRefreshToken(String token) {
        this.refreshToken = token;
    }

    public void setStatusType(String status) {
        this.statusType = status;
    }

    public void setStatusDescription(String desc) {
        this.statusDescription = desc;
    }

    public void setTimeUntil(String time) {
        this.timeUntil = time;
    }

    public String getStatusType() {
        return this.statusType;
    }

    public String getStatusDescription() {
        return this.statusDescription;
    }

    public String getTimeUntil() {
        return this.timeUntil;
    }
}
